the-collab-lab/tcl-59-smart-shopping-list
# Creating a search box that allows the user to filter their list items

## Description

The code we have added was to add a search box to the list view that allows the user to filter their shopping list to narrow down their list. The list filters as the user types, and the user can search by any part of the item name, rather than simply the beginning. 

There is an X to clear the field and I (Karen 😄) learnt from Mike that adding a type="search" attribute automatically gives us this functionality. 

The code to render the search bar itself is in List.jsx. The code to filter the list items is in App.jsx, and works by creating some searchQuery state. We pass down setSearchQuery into List.jsx, and set it as e.target.value as the user types into the search field. We also added a filterList function that when called sees whether the item's name includes what's typed into the search field. This returns an updated list that we pass down as 'data' to be rendered in List.jsx.  

We used the string method .toLowerCase because .includes is case sensitive.

## Related Issue

closes issue https://github.com/the-collab-lab/tcl-59-smart-shopping-list/issues/6

## Acceptance Criteria


- A text field (with semantic `<label>`) which narrows down the list as the user types
- When there’s text in the field, some kind of button (e.g., an X) to clear the field. When the field is cleared, the list is reset to its unfiltered state.

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### After

<img width="431" alt="image" src="https://user-images.githubusercontent.com/97447750/233388605-b5e3c860-c464-45df-a65a-32767a537e2f.png">

## Testing Steps / QA Criteria

Please pull this branch down with the command- `git pull origin mo-kp-filter-list-items` and check that branch out with git checkout `mo-kp-filter-list-items`. 

**You'll need to create a list, and then add a few items before you test this, unless you have a token in your localStorage already with some items associated with it! You could also temporarily set the listToken back to 'my test list' as that now has a very lengthy shopping list associated with it so it's quite good for testing the search**

Then, in List view, please first check you can see the full list of items, and then that, as you type, you see the list dynamically filter. Try any part of the name not just the start and please try the X too.
